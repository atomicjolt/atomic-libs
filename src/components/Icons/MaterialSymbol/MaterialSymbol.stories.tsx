import { Meta, StoryObj } from "@storybook/react";
import { MaterialSymbol } from ".";

const meta: Meta<typeof MaterialSymbol> = {
  title: "Icons/MaterialSymbol",
  component: MaterialSymbol,
  parameters: {
    layout: "centered",
  },
};

export default meta;

type Story = StoryObj<typeof MaterialSymbol>;

export const Primary: Story = {
  args: {
    symbol: "label",
    size: "large",
    variant: "outlined",
  },
};
