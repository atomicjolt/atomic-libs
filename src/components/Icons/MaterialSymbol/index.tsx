import React from "react";
import cn from "classnames";
import { HasClassName, MaterialIcons } from "../../../types";
import { useVariantClass } from "../../../hooks";
import { StyledIcon } from "../Icons.styles";

export type MaterialSymbolVariants = "outlined" | "rounded" | "sharp";

export interface MaterialSymbolProps
  extends Omit<
      React.HTMLProps<HTMLElement>,
      "size" | "className" | "ref" | "as"
    >,
    HasClassName {
  symbol: MaterialIcons;
  /** The type of material symbol to
   * render. Note that the font for that style needs to
   * be in scope for it to render properly*/
  variant?: MaterialSymbolVariants;
  size?: "small" | "medium" | "large";
  disabled?: boolean;
}

/** Render out material-symbols with sensible defaults. No Material
 * Symbol Font variant is loaded when using <LoadFonts /> by default,
 * so you need to include it explictly, or load it another way */
export default function MaterialSymbol(props: MaterialSymbolProps) {
  const {
    symbol,
    className,
    variant = "outlined",
    size = "medium",
    disabled = false,
    ...rest
  } = props;

  const variantClass = useVariantClass("material-symbols", variant, "-");

  return (
    <StyledIcon
      className={cn(variantClass, className, `is-${size}`, {
        "is-disabled": disabled,
      })}
      aria-hidden={true}
      {...rest}
    >
      {symbol}
    </StyledIcon>
  );
}
