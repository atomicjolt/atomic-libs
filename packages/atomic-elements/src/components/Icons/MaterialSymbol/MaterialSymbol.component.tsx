import React from "react";
import { useFocusable } from "@react-aria/focus";
import { filterDOMProps, mergeProps, useObjectRef } from "@react-aria/utils";
import {
  IconComponentBase,
  MaterialSymbolFontSettingsOptions,
  MaterialSymbols,
  MaterialSymbolVariants,
} from "../../../types";
import { useRenderProps } from "@hooks/useRenderProps";
import { useMaterialSymbolFontVariationSettings } from "@hooks/useMaterialSymbolFontVariationSettings";
import { StyledIcon } from "../Icons.styles";

export interface MaterialSymbolProps
  extends IconComponentBase<never>,
    MaterialSymbolFontSettingsOptions {
  symbol: MaterialSymbols;
  /** The type of material symbol to
   * render. Note that the font for that style needs to
   * be in scope for it to render properly */
  variant?: MaterialSymbolVariants;
}

/** Render out material-symbols with sensible defaults. */
export const MaterialSymbol = React.forwardRef<
  HTMLElement,
  MaterialSymbolProps
>(function MaterialSymbol(props, forwardedRef) {
  const {
    symbol,
    className,
    variant = "outlined",
    size = "medium",
    isDisabled = false,
    isFocusable = false,
    style,
    weight,
    grade,
    opticalSize,
    fill,
    ...rest
  } = props;

  const ref = useObjectRef(forwardedRef);

  const materialSymbolClass = `material-symbols-${variant}`;

  const renderProps = useRenderProps({
    componentClassName: "aje-icon",
    className: [materialSymbolClass, className],
    size,
    style,
    selectors: {
      "data-disabled": isDisabled,
    },
  });

  const fontVariationSettings = useMaterialSymbolFontVariationSettings({
    fill,
    weight,
    grade,
    opticalSize,
  });

  const componentProps = [filterDOMProps(rest), renderProps];

  const { focusableProps } = useFocusable({}, ref);

  if (isFocusable) {
    componentProps.push(focusableProps);
  }

  const mergedProps = mergeProps(...componentProps);

  return (
    <StyledIcon
      ref={ref}
      aria-hidden
      {...mergedProps}
      style={{
        fontVariationSettings: fontVariationSettings || undefined,
        ...mergedProps.style,
      }}
    >
      {symbol}
    </StyledIcon>
  );
});
