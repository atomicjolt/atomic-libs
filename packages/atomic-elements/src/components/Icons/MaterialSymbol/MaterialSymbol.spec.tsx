import { describe, it, expect } from "vitest";
import { render, screen } from "@testing-library/react";
import { MaterialSymbol } from "./MaterialSymbol.component";

describe("MaterialSymbol", () => {
  it("renders correctly", () => {
    const { container } = render(<MaterialSymbol symbol="home" />);
    expect(container).toMatchSnapshot();
  });

  it("should not be focusable by default", () => {
    render(<MaterialSymbol symbol="home" />);
    const icon = screen.getByText("home");
    expect(icon.tabIndex).toBe(-1);
  });

  it("should be focusable when isFocusable is true", () => {
    render(<MaterialSymbol symbol="home" isFocusable />);
    const icon = screen.getByText("home");
    expect(icon.tabIndex).toBe(0);
  });
});
