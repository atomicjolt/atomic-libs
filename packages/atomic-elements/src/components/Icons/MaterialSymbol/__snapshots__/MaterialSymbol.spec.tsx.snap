// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MaterialSymbol > renders correctly 1`] = `
<div>
  <i
    aria-hidden="true"
    class="sc-FEMpB jyuubH aje-icon material-symbols-outlined is-medium"
  >
    home
  </i>
</div>
`;
